uaclient/event_logger.py

```diff
@@ -0,0 +1,135 @@
+those events in real time or through a machine-readable format.
+"""
+
+import enum
+import json
+from typing import Dict, List, Optional, Set  # noqa: F401
+
+_event_logger = None
+
+
+def get_event_logger():
+    global _event_logger
+
+    if _event_logger is None:
+        _event_logger = EventLogger()
+
+    return _event_logger
+
+
+@enum.unique
+class EventLoggerMode(enum.Enum):
+    """
+    Defines event logger supported modes.
+    Currently, we only support the cli and machine-readable
+    mode. On cli mode, we will print to stdout/stderr any
+    event that we receive. On machine-readable mode, we will
+    store those events and parse them for an specified
+    format.
+    """
+
+    CLI = object()
+    MACHINE_READABLE = object()
+
+
+class EventLogger:
+    def __init__(self):
+        self._error_events = []  # type: List[Dict[str, Optional[str]]]
+        self._warning_events = []  # type: List[Dict[str, Optional[str]]]
+        self._processed_services = []  # type: List[str]
+        self._failed_services = set()  # type: Set[str]
+        self._needs_reboot = False
+
+        # By default, the event logger will be on CLI mode,
+        # printing every event it receives.
+        self._event_logger_mode = EventLoggerMode.CLI
+
+    def reset(self):
+        self._error_events = []
+        self._warning_events = []
+        self._processed_services = []
+        self._failed_services = set()
+        self._needs_reboot = False
+
+    def set_event_mode(self, event_mode: EventLoggerMode):
+        self._event_logger_mode = event_mode
+
+    def info(self, info_msg: str):
+        if self._event_logger_mode == EventLoggerMode.CLI:
+            print(info_msg)
+
+    def _record_dict_event(
+        self,
+        msg: str,
+        service: Optional[str],
+        event_dict: List[Dict[str, Optional[str]]],
+    ):
+        event_type = "service" if service else "system"
+        event_dict.append(
+            {"type": event_type, "service": service, "message": msg}
+        )
+
+    def error(self, error_msg: str, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.MACHINE_READABLE:
+            self._record_dict_event(
+                msg=error_msg, service=service, event_dict=self._error_events
+            )
+
+    def warning(self, warning_msg: str, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.MACHINE_READABLE:
+            self._record_dict_event(
+                msg=warning_msg,
+                service=service,
+                event_dict=self._warning_events,
+            )
+
+    def raise_error(self, exception: Exception, service: Optional[str] = None):
+        if self._event_logger_mode == EventLoggerMode.CLI:
+            raise exception
+        else:
+            error_msg = getattr(exception, "msg") or str(exception)
+            self.error(error_msg=error_msg, service=service)
```

Comment:
Why would we only `raise` if in cli mode? I would think `json` mode shouldn't alter behavior in that way

Reply:
Thinking about this more (and assuming `cli` and `json` mode should `raise` in the same situations). I think I'd expect that `raise` shouldn't happen at all in this module. This module can just record the error and the caller can raise if necessary. What do you think?

Reply:
You mean on the cli function, we would report the error to the event logger and raise it as well ? If yes, my concern is that during the json mode, we could incorrectly print an exception to the user.

Or do you mean that on cli, we should check the format type before raising the exception ?

Reply:
My main worry is generally that in a given situation, we should either always `raise` or never `raise`. We shouldn't `raise` or not depending on whether the output is going to be cli or json. My reason being: it is hard to tell by reading the code if they actually do the same thing or if, because it doesn't `raise` during json mode, then json mode will actually keep executing further than it would have in cli mode.

I think it should be safe to use the event module in the main_error_handler because at that point the mode should've been set.